client/service/src/chain_ops.rs

```diff
@@ -116,55 +295,32 @@ impl<
 			let client = &self.client;
 			let queue = &mut self.import_queue;
 
-			// Start by reading the number of blocks if not done so already.
-			let count = match count {
-				Some(c) => c,
-				None => {
-					let c: u64 = match Decode::decode(&mut io_reader_input) {
-						Ok(c) => c,
-						Err(err) => {
-							let err = format!("Error reading file: {}", err);
-							return std::task::Poll::Ready(Err(From::from(err)));
-						},
-					};
-					info!("📦 Importing {} blocks", c);
-					count = Some(c);
-					c
-				}
-			};
-
 			// Read blocks from the input.
-			if read_block_count < count {
-				match SignedBlock::<Self::Block>::decode(&mut io_reader_input) {
-					Ok(signed) => {
-						let (header, extrinsics) = signed.block.deconstruct();
-						let hash = header.hash();
-						// import queue handles verification and importing it into the client
-						queue.import_blocks(BlockOrigin::File, vec![
-							IncomingBlock::<Self::Block> {
-								hash,
-								header: Some(header),
-								body: Some(extrinsics),
-								justification: signed.justification,
-								origin: None,
-								allow_missing_state: false,
-								import_existing: force,
-							}
-						]);
+			match Pin::new(&mut block_stream).poll_next(cx) {
+				Poll::Ready(None) => {
+					let read_block_count = block_stream.read_block_count();
+					let count = block_stream.count();
+
+					if importing_is_done(count, read_block_count, link.imported_blocks) {
+						info!(
+							"🎉 Imported {} blocks. Best: #{}",
+							read_block_count, client.chain_info().best_number
+						);
+						return std::task::Poll::Ready(Ok(()))
 					}
-					Err(e) => {
-						warn!("Error reading block data at {}: {}", read_block_count, e);
-						return std::task::Poll::Ready(Ok(()));
+				},
+				Poll::Ready(Some(block_result)) => {
+					match block_result {
+						Ok(signed_block) => import_block_to_queue(signed_block, queue, force),
```

Comment:
Here we need to check how many blocks are already in the queue and wait if there's more than say 1024

Reply:
Good point. I tried making this in https://github.com/paritytech/substrate/pull/5816/commits/17f4c74e8fcd3416067b137c79f93c0885097afa, please let me know what you think! :)